device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/DeviceIO.java

```diff
@@ -429,4 +432,33 @@ public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeC
         //Codes_SRS_DEVICE_IO_34_020: [This function shall register the callback with the transport.]
         this.transport.registerConnectionStatusChangeCallback(statusChangeCallback, callbackContext);
     }
+
+    /*
+     * IotHubTransport layer will notify this layer when the connection is established and when it is lost. This layer should start/stop
+     * the send/receive threads accordingly
+     */
+    @Override
+    public void execute(IotHubConnectionStatus status, IotHubConnectionStatusChangeReason statusChangeReason, Throwable throwable, Object callbackContext)
+    {
+        if (status == IotHubConnectionStatus.DISCONNECTED || status == IotHubConnectionStatus.DISCONNECTED_RETRYING)
+        {
+            // No need to keep spawning send/receive tasks during reconnection or when the client is closed
+            if (this.sendTaskScheduler != null)
+            {
+                this.sendTaskScheduler.shutdown();
+            }
+
+            if (this.receiveTaskScheduler != null)
+            {
+                this.receiveTaskScheduler.shutdown();
+            }
+        }
+        else if (status == IotHubConnectionStatus.CONNECTED)
+        {
+            // Restart the task scheduler so that send/receive tasks start spawning again
+            this.startWorkerThreads();
+        }
+
+        this.state = status;
```

Comment:
shouldn't these status updates be made over a lock?

Reply:
Sure, I'll protect this with the same lock that I'll add for open/close